### DISCUSSION TITLE
What was r2 supposed to be in scalar static test vectors?

### DISCUSSION THREAD

**Discussion by Coding-Enthusiast** [bitcoin-core/secp256k1#1463] [Q&A] [answered]
Scalar has a bunch of static test vectors and there is an `r2` value defined in the hard-coded values (bytes) that is never used. What is/was its purpose?  

https://github.com/bitcoin-core/secp256k1/blob/77af1da9f631fa622fb5b5895fd27be431432368/src/tests.c#L2927

**Comment by jonasnick** [comment, marked as answer]
`r2` was used to test the correctness of `scalar_sqr`. The tests and `scalar_sqr` were removed in commit [5437e7bdfbffddf69fdf7b4af7e997c78f5dafbf](https://github.com/bitcoin-core/secp256k1/commit/5437e7bdfbffddf69fdf7b4af7e997c78f5dafbf#diff-c2d5f1f7616875ab71cd41b053cfb428696988ff89642b931a0963d50f34f7e8L2195). You are right that it is unused right now.

I don't know the exact intentions of the tests, but there's also the possibility that the `r2` static test vectors also increased test coverage of `mul` because the removed lines also essentially tested this:
```C
            secp256k1_scalar_mul(&z, &x, &x);
            CHECK(secp256k1_scalar_eq(&r2, &z));
```

We could either remove `r2` and `res[i][1]` or use them to test `mul`.

**Reply by real-or-random** [reply]
> We could either remove `r2`

The quote line `secp256k1_scalar_set_b32(&r2, res[i][1], &overflow);` also sets `overflow`, which is checked in the next line. If we removed `r2`, we would also need to remove that check.

>  use them to test `mul`.

Sounds better.
